File: WexbimHarness/WexBimStream.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Runtime.Serialization;
using System.Text;
using System.Threading.Tasks;
using Xbim.Common.Geometry;
using Xbim.Ifc;

namespace AimViewModels.Shared.Helpers
{

    public class WexBimHeader
    {
        public WexBimHeader()
        {
            MagicNumber = IfcStore.WexBimId;
            Version = 3; //modified to support regions correctly
            
        }
        public Int32 MagicNumber { get; private set; }
        public byte Version { get; private set; }
        public Int32 ShapeCount { get; set; }
        public Int32 VertexCount { get; set; }
        public Int32 TriangleCount { get; set; }
        public Int32 MatrixCount { get; set; }
        public Int32 ProductCount { get; set; }
        public Int32 StyleCount { get; set; }
        public Single OneMeter { get; set; }
        public Int16 RegionCount { get; set; }
        public void WriteToStream(BinaryWriter writer)
        {
            writer.Write(MagicNumber);
            writer.Write(Version);
            writer.Write(ShapeCount);
            writer.Write(VertexCount);
            writer.Write(TriangleCount);
            writer.Write(MatrixCount);
            writer.Write(ProductCount);
            writer.Write(StyleCount);
            writer.Write(OneMeter);
            writer.Write(RegionCount);
        }

        public static WexBimHeader ReadFromStream(BinaryReader reader)
        {
            var header = new WexBimHeader();
            header.MagicNumber = reader.ReadInt32();
            header.Version = reader.ReadByte();
            header.ShapeCount = reader.ReadInt32();
            header.VertexCount = reader.ReadInt32();
            header.TriangleCount = reader.ReadInt32();
            header.MatrixCount = reader.ReadInt32();
            header.ProductCount = reader.ReadInt32();
            header.StyleCount = reader.ReadInt32();
            header.OneMeter = reader.ReadSingle();
            header.RegionCount = reader.ReadInt16();
            return header;
        }

    }
    public class WexBimRegion
    {
        public WexBimRegion()
        {
            _geometryModels = new List<WexBimGeometryModel>();
        }
        public Int32 Population { get; set; }
        public Single CentreX { get; set; }
        public Single CentreY { get; set; }
        public Single CentreZ { get; set; }
        public XbimRect3D BoundingBox;
        private List<WexBimGeometryModel> _geometryModels;
        public IList<WexBimGeometryModel> GeometryModels => _geometryModels;

        public int MatrixCount()
        {
            return _geometryModels.Sum(gm=>gm.MatrixCount());
        }

        public int ShapeCount()
        {
            return _geometryModels.Sum(gm => gm.ShapeCount());
        }
        public int TriangleCount()
        {
            return _geometryModels.Sum(gm => gm.TriangleCount());
        }
        public int VertexCount()
        {
            return _geometryModels.Sum(gm => gm.VertexCount());
        }
        public void AddGeometryModel(byte[] triangulation, IEnumerable<WexBimShapeMultiInstance> instances)
        {
            var gm = new WexBimGeometryModel { Geometry = new WexBimMesh(triangulation) };
            gm.AddInstances(instances);
            _geometryModels.Add(gm);
        }
        public void AddGeometryModel(byte[] triangulation, WexBimShapeSingleInstance singleInstance)
        {
            var gm = new WexBimGeometryModel { Geometry = new WexBimMesh(triangulation) };
            gm.AddInstance(singleInstance);
            _geometryModels.Add(gm);
        }

        public void AddGeometryModel(WexBimGeometryModel gm)
        {
            _geometryModels.Add(gm);
        }
        public void WriteToStream(BinaryWriter writer)
        {
           
            writer.Write(Population);
            writer.Write(CentreX);
            writer.Write(CentreY);
            writer.Write(CentreZ);
            writer.Write(BoundingBox.ToFloatArray());
        }
        public static WexBimRegion ReadFromStream(BinaryReader reader)
        {
            var region = new WexBimRegion();
            region.Population = reader.ReadInt32();
            region.CentreX = reader.ReadSingle();
            region.CentreY = reader.ReadSingle();
            region.CentreZ = reader.ReadSingle();
            region.BoundingBox = XbimRect3D.FromArray(reader.ReadBytes(6*sizeof(float)));        
            return region;
        }

        
    }
    public class WexBimStyle
    {
        public Int32 StyleId { get; set; }
        public Single Red { get; set; }
        public Single Green { get; set; }
        public Single Blue { get; set; }
        public Single Alpha { get; set; }
        public void WriteToStream(BinaryWriter writer)
        {
            writer.Write(StyleId);
            writer.Write(Red);
            writer.Write(Green);
            writer.Write(Blue);
            writer.Write(Alpha);
        }

        public static WexBimStyle ReadFromStream(BinaryReader reader)
        {
            var style = new WexBimStyle();
            style.StyleId = reader.ReadInt32();
            style.Red = reader.ReadSingle();
            style.Green = reader.ReadSingle();
            style.Blue = reader.ReadSingle();
            style.Alpha = reader.ReadSingle();
            return style;
        }
    }
    public class WexBimProduct
    {
        public Int32 ProductLabel { get; set; }
        public Int16 ProductType { get; set; }
        public XbimRect3D BoundingBox{ get; set; }
        public void WriteToStream(BinaryWriter writer)
        {
            
            writer.Write(ProductLabel);
            writer.Write(ProductType);
            writer.Write(BoundingBox.ToFloatArray());
        }

        public static WexBimProduct ReadFromStream(BinaryReader reader)
        {
            var product = new WexBimProduct();
            product.ProductLabel = reader.ReadInt32();
            product.ProductType = reader.ReadInt16();
            product.BoundingBox = XbimRect3D.FromArray(reader.ReadBytes(6 * sizeof(float)));
            return product;
        }
    }
    public interface IWexBimShape
    {
        bool IsSingleInstance { get; }
        Int32 ProductLabel { get; set; }
        Int16 InstanceTypeId { get; set; }
        Int32 InstanceLabel { get; set; }
        Int32 StyleId { get; set; }
        byte[] Transformation { get; set; }
        void WriteToStream(BinaryWriter writer);
    }
    public class WexBimShapeSingleInstance : IWexBimShape
    {
        public Int32 ProductLabel { get; set; }
        public Int16 InstanceTypeId { get; set; }
        public Int32 InstanceLabel { get; set; }
        public Int32 StyleId { get; set; }

        public bool IsSingleInstance
        {
            get
            {
                return true;
            }
        }

        public byte[] Transformation
        {
            get
            {
                return null;
            }

            set
            {
                throw new NotImplementedException();
            }
        }

        public virtual void WriteToStream(BinaryWriter writer)
        {
            writer.Write(ProductLabel);
            writer.Write(InstanceTypeId);
            writer.Write(InstanceLabel);
            writer.Write(StyleId);
        }

        public static WexBimShapeSingleInstance ReadFromStream(BinaryReader reader)
        {
            var shape = new WexBimShapeSingleInstance();
            shape.ProductLabel = reader.ReadInt32();
            shape.InstanceTypeId = reader.ReadInt16();
            shape.InstanceLabel = reader.ReadInt32();
            shape.StyleId = reader.ReadInt32();
            return shape;
        }
    }

    /// <summary>
    /// Special kind of shape header for multiple instance shapes which has a transform
    /// </summary>
    public class WexBimShapeMultiInstance :  IWexBimShape
    {
        public Int32 ProductLabel { get; set; }
        public Int16 InstanceTypeId { get; set; }
        public Int32 InstanceLabel { get; set; }
        public Int32 StyleId { get; set; }
        public byte[] Transformation { get; set; }

        public bool IsSingleInstance
        {
            get
            {
               return false;
            }
        }

        public void WriteToStream(BinaryWriter writer)
        {
            writer.Write(ProductLabel);
            writer.Write(InstanceTypeId);
            writer.Write(InstanceLabel);
            writer.Write(StyleId);
            Debug.Assert(Transformation.Length == 16 * sizeof(double));
            writer.Write(Transformation);          
        }

        public static WexBimShapeMultiInstance ReadFromStream(BinaryReader reader)
        {
            var shape = new WexBimShapeMultiInstance();
            shape.ProductLabel = reader.ReadInt32();
            shape.InstanceTypeId = reader.ReadInt16();
            shape.InstanceLabel = reader.ReadInt32();
            shape.StyleId = reader.ReadInt32();
            shape.Transformation=reader.ReadBytes(16 * sizeof(double));
            return shape;
        }
    }

    public class WexBimGeometryModel
    {
        private List<IWexBimShape> _shapes;

        public WexBimGeometryModel()
        {
            _shapes = new List<IWexBimShape>();
        }
        public IEnumerable<IWexBimShape> Shapes => _shapes;
        public WexBimMesh Geometry { get; set; }

        public int MatrixCount()
        {
            return _shapes.OfType<WexBimShapeMultiInstance>().Count();
        }

        public int ShapeCount()
        {
            return _shapes.Count;
        }

        public int TriangleCount()
        {
            return Geometry.TriangleCount * _shapes.Count;
        }
        public int VertexCount()
        {
            return Geometry.VertexCount;
        }
        public void AddInstances(IEnumerable<WexBimShapeMultiInstance> instances)
        {
            _shapes.AddRange(instances);
        }

        public void AddInstance(WexBimShapeSingleInstance singleInstance)
        {
            _shapes.Add(singleInstance);
        }
        public void WriteToStream(BinaryWriter writer)
        {
            writer.Write(_shapes.Count);
            foreach (var shape in _shapes) //does matter which one it should be fine
            {
                shape.WriteToStream(writer);
            }
            writer.Write(Geometry.Length);
            writer.Write(Geometry.ToByteArray());
        }

        public static WexBimGeometryModel ReadFromStream(BinaryReader reader)
        {
            var geometry = new WexBimGeometryModel();
            var numShapes = reader.ReadInt32();
            if(numShapes>1) //we have a multi used geometry
            {
                for (int i = 0; i < numShapes; i++)
                {
                    geometry._shapes.Add(WexBimShapeMultiInstance.ReadFromStream(reader));
                }
            }
            else //just the one
            {
                geometry._shapes.Add( WexBimShapeSingleInstance.ReadFromStream(reader));
            }
            //read the geometry
            var numBytes = reader.ReadInt32();
            geometry.Geometry = new WexBimMesh(reader.ReadBytes(numBytes));
            return geometry;
        }

      
    }
    public class WexBimStream
    {
        public WexBimStream()
        {
            Header = new WexBimHeader();
            _regions = new List<WexBimRegion>();
            _styles = new List<WexBimStyle>();
            _products = new Dictionary<int, WexBimProduct>();          
        }
        public WexBimHeader Header { get; private set; }
        private List<WexBimRegion> _regions;
        private List<WexBimStyle> _styles;
        private Dictionary<int, WexBimProduct> _products;
       

        public IList<WexBimRegion> Regions => _regions;
        public IList<WexBimStyle> Styles => _styles;
        public IEnumerable<WexBimProduct> Products => _products.Values;
       
        public void AddRegion(WexBimRegion region)
        {
            _regions.Add(region);
        }
        public void AddStyle(WexBimStyle style)
        {
            _styles.Add(style);
        }
       
        /// <summary>
        /// Adds a product part representation to the view, if parts of the product are already added the bounding box is expanded, the first product type is retained
        /// </summary>
        /// <param name="product"></param>
        public void AddProduct(WexBimProduct product)
        {
            WexBimProduct existingProduct;
            if (_products.TryGetValue(product.ProductLabel, out existingProduct))
            {
                var bb = existingProduct.BoundingBox;
                bb.Union(product.BoundingBox);
                existingProduct.BoundingBox = bb;
            }
            else
                _products.Add(product.ProductLabel, product);
        }

        public void WriteToStream(BinaryWriter writer)
        {
            //first write the header but rememeber the pont on the stram as we have to overrite when all the data is known
            var startPosition = writer.Seek(0, SeekOrigin.Current);
            //set up the header
            UpdateHeader();
            Header.WriteToStream(writer);
            foreach (var region in Regions) region.WriteToStream(writer);
            foreach (var style in Styles) style.WriteToStream(writer);
            foreach (var product in Products) product.WriteToStream(writer);
            foreach (var region in Regions)
            {
                writer.Write(region.GeometryModels.Count);
                foreach (var geometry in region.GeometryModels) geometry.WriteToStream(writer);
            }
        }

        static public WexBimStream ReadFromStream(BinaryReader reader)
        {
            var wexBimStream = new WexBimStream();
            wexBimStream.Header = WexBimHeader.ReadFromStream(reader);
            for (int i = 0; i < wexBimStream.Header.RegionCount; i++)
                wexBimStream.AddRegion(WexBimRegion.ReadFromStream(reader));
            for (int i = 0; i < wexBimStream.Header.StyleCount; i++)
                wexBimStream.AddStyle(WexBimStyle.ReadFromStream(reader));
            for (int i = 0; i < wexBimStream.Header.ProductCount; i++)
                wexBimStream.AddProduct(WexBimProduct.ReadFromStream(reader));
            foreach (var region in wexBimStream.Regions)
            {
                var geometryCount = reader.ReadInt32();
                for (int i = 0; i < geometryCount; i++)
                    region.AddGeometryModel(WexBimGeometryModel.ReadFromStream(reader));
            }
            return wexBimStream;
        }
        public void UpdateHeader()
        {
            Header.MatrixCount = Regions.Sum(r => r.MatrixCount());
            Header.ProductCount = _products.Count;
            Header.RegionCount = (short)_regions.Count;
            Header.ShapeCount = Regions.Sum(r => r.ShapeCount());
            Header.StyleCount = _styles.Count;
            Header.TriangleCount= Regions.Sum(r => r.TriangleCount());
            Header.VertexCount = Regions.Sum(r => r.VertexCount());
        }
        /// <summary>
        /// Reads an entire WexBim stream
        /// </summary>
        /// <param name="reader"></param>
        /// <returns></returns>
        //public static WexBimHelper ReadFromStream(BinaryReader reader)
        //{

        //}



    }
}

File: src/stream/wexbim-stream.ts
import { BinaryReader } from "../reader/binary-reader";
import { ProductType } from "../product-type";
import { WexBimMesh } from "./wexbim-mesh";

export class WexBimHeader {
    constructor() {
        this.Version = 3; //modified to support regions correctly

    }
    public static MagicNumber: number = 94132117;
    public Version: number;
    public ShapeCount: number;
    public VertexCount: number;
    public TriangleCount: number;
    public MatrixCount: number;
    public ProductCount: number;
    public StyleCount: number;
    public OneMeter: number;
    public RegionCount: number;


    public static ReadFromStream(reader: BinaryReader): WexBimHeader {
        var header = new WexBimHeader();
        let magicNumber = reader.readInt32();
        if (magicNumber !== WexBimHeader.MagicNumber) {
            throw new Error("This is not a valid wexbim file. Magic number mismatch.");
        }

        header.Version = reader.readByte();
        header.ShapeCount = reader.readInt32();
        header.VertexCount = reader.readInt32();
        header.TriangleCount = reader.readInt32();
        header.MatrixCount = reader.readInt32();
        header.ProductCount = reader.readInt32();
        header.StyleCount = reader.readInt32();
        header.OneMeter = reader.readFloat32();
        header.RegionCount = reader.readInt16();
        return header;
    }

}

// tslint:disable: max-classes-per-file
export class WexBimRegion {

    public Population: number;
    public Centre: Float32Array;
    public BoundingBox: Float32Array;
    public GeometryModels: WexBimGeometryModel[] = new Array<WexBimGeometryModel>();

    public MatrixCount(): number {
        var count = 0;
        this.GeometryModels.forEach((gm) => count += gm.MatrixCount);
        return count;
    }

    public ShapeCount(): number {
        var count = 0;
        this.GeometryModels.forEach((gm) => count += gm.ShapeCount);
        return count;
    }
    public TriangleCount(): number {
        var count = 0;
        this.GeometryModels.forEach((gm) => count += gm.TriangleCount);
        return count;
    }

    public VertexCount(): number {
        var count = 0;
        this.GeometryModels.forEach((gm) => count += gm.VertexCount);
        return count;
    }

    public AddGeometryModel(model: WexBimGeometryModel): void {
        this.GeometryModels.push(model);
    }

    public static ReadFromStream(reader: BinaryReader): WexBimRegion {
        var region = new WexBimRegion();
        region.Population = reader.readInt32();
        region.Centre = reader.readFloat32Array(3);
        region.BoundingBox = reader.readFloat32Array(6);
        return region;
    }
}

export class WexBimStyle {
    public StyleId: number;
    public RGBA: Float32Array;

    public static ReadFromStream(reader: BinaryReader): WexBimStyle {
        var style = new WexBimStyle();
        style.StyleId = reader.readInt32();
        style.RGBA = reader.readFloat32Array(4);
        return style;
    }
}

export class WexBimProduct {
    public ProductLabel: number;
    public ProductType: ProductType;
    public BoundingBox: Float32Array;

    //inverse, filled in at the end of parsing
    public Shapes: IWexBimShape[];

    public static ReadFromStream(reader: BinaryReader): WexBimProduct {
        var product = new WexBimProduct();
        product.ProductLabel = reader.readInt32();
        product.ProductType = reader.readInt16();
        product.BoundingBox = reader.readFloat32Array(6);
        return product;
    }
}

export interface IWexBimShape {
    IsSingleInstance: boolean;
    ProductLabel: number;
    InstanceTypeId: number;
    InstanceLabel: number;
    StyleId: number;
    Transformation: Float64Array;

    //inverse relation
    GeometryModel: WexBimGeometryModel;
}

export class WexBimShapeSingleInstance implements IWexBimShape {
    public IsSingleInstance: boolean = true;
    public ProductLabel: number;
    public InstanceTypeId: number;
    public InstanceLabel: number;
    public StyleId: number;
    public Transformation: Float64Array = null;

    //inverse relation
    public GeometryModel: WexBimGeometryModel;

    public static ReadFromStream(reader: BinaryReader, model: WexBimGeometryModel): WexBimShapeSingleInstance {
        var shape = new WexBimShapeSingleInstance();
        shape.ProductLabel = reader.readInt32();
        shape.InstanceTypeId = reader.readInt16();
        shape.InstanceLabel = reader.readInt32();
        shape.StyleId = reader.readInt32();

        shape.GeometryModel = model;
        return shape;
    }
}

/// <summary>
/// Special kind of shape header for multiple instance shapes which has a transform
/// </summary>
export class WexBimShapeMultiInstance implements IWexBimShape {
    public IsSingleInstance: boolean = false;
    public ProductLabel: number;
    public InstanceTypeId: number;
    public InstanceLabel: number;
    public StyleId: number;
    public Transformation: Float64Array = null;

    //inverse relation
    public GeometryModel: WexBimGeometryModel;

    public static ReadFromStream(reader: BinaryReader, model: WexBimGeometryModel): WexBimShapeMultiInstance {
        var shape = new WexBimShapeMultiInstance();
        shape.ProductLabel = reader.readInt32();
        shape.InstanceTypeId = reader.readInt16();
        shape.InstanceLabel = reader.readInt32();
        shape.StyleId = reader.readInt32();
        shape.Transformation = reader.readFloat64Array(16);

        //set inverse
        shape.GeometryModel = model;
        return shape;
    }
}

export class WexBimGeometryModel {

    public Shapes: IWexBimShape[] = new Array<IWexBimShape>();
    public Geometry: WexBimMesh;

    public get MatrixCount(): number {
        return this.Shapes.filter((s) => s instanceof WexBimShapeMultiInstance).length;
    }

    public get ShapeCount(): number {
        return this.Shapes.length;
    }

    public get TriangleCount(): number {
        return this.Geometry.TriangleCount;
    }
    public get VertexCount(): number {
        return this.Geometry.VertexCount;
    }

    public static ReadFromStream(reader: BinaryReader): WexBimGeometryModel {
        var geometry = new WexBimGeometryModel();
        var numShapes = reader.readInt32();
        if (numShapes > 1) { //we have a multi used geometry

            for (let i = 0; i < numShapes; i++) {
                var instance = WexBimShapeMultiInstance.ReadFromStream(reader, geometry);
                geometry.Shapes[instance.ProductLabel] = instance;
            }
        } else { //just the one
            var instance = WexBimShapeSingleInstance.ReadFromStream(reader, geometry);
            geometry.Shapes[instance.ProductLabel] = instance;
        }
        //read the geometry
        let numBytes = reader.readInt32();
        geometry.Geometry = new WexBimMesh(reader.readData(numBytes));
        return geometry;
    }


}
export class WexBimStream {
    public Header: WexBimHeader;
    public Regions: WexBimRegion[] = new Array<WexBimRegion>();
    public Styles: WexBimStyle[] = new Array<WexBimStyle>();
    public Products: { [id: number]: WexBimProduct } = {};

    public AddRegion(region: WexBimRegion): void {
        this.Regions.push(region);
    }

    public AddStyle(style: WexBimStyle): void {
        this.Styles.push(style);
    }

    private Union(boxA: Float32Array, boxB: Float32Array): Float32Array {
        var result = new Float32Array(6);
        if (boxA == null && boxB != null) {
            result.set(boxB);
            return result;
        }

        if (boxB == null) {
            return null;
        }

        result[0] = Math.min(boxA[0], boxB[0]);
        result[1] = Math.min(boxA[1], boxB[1]);
        result[2] = Math.min(boxA[2], boxB[2]);

        result[3] = Math.max((boxA[0] + boxA[3]), (boxB[0] + boxB[3])) - result[0];
        result[4] = Math.max((boxA[1] + boxA[4]), (boxB[1] + boxB[4])) - result[1];
        result[5] = Math.max((boxA[2] + boxA[5]), (boxB[2] + boxB[5])) - result[2];

        return result;
    }

    /// <summary>
    /// Adds a product part representation to the view, if parts of the product are already added the bounding box is expanded, the first product type is retained
    /// </summary>
    /// <param name="product"></param>
    public AddProduct(product: WexBimProduct): void {
        let existingProduct = this.Products[product.ProductLabel];

        if (existingProduct) {
            existingProduct.BoundingBox = this.Union(existingProduct.BoundingBox, product.BoundingBox);
        } else {
            this.Products[product.ProductLabel] = product;
        }
    }

    public static Load(source: string | Blob | File | ArrayBuffer, callback: (wexbim: WexBimStream) => void): void {
        if (source == null) {
            throw new Error("Undefined source");
        }

        if (callback == null) {
            throw new Error("You have to use callback to get the stream");
        }

        var reader = new BinaryReader();
        reader.onloaded = (r) => {
            let wexbim = WexBimStream.ReadFromStream(r);
            if (callback) {
                callback(wexbim);
            }
        };
        reader.load(source, null, null);
    }

    public static ReadFromStream(reader: BinaryReader): WexBimStream {
        var wexBimStream = new WexBimStream();
        wexBimStream.Header = WexBimHeader.ReadFromStream(reader);

        for (let i = 0; i < wexBimStream.Header.RegionCount; i++) {
            wexBimStream.AddRegion(WexBimRegion.ReadFromStream(reader));
        }

        for (let i = 0; i < wexBimStream.Header.StyleCount; i++) {
            wexBimStream.AddStyle(WexBimStyle.ReadFromStream(reader));
        }

        for (let i = 0; i < wexBimStream.Header.ProductCount; i++) {
            wexBimStream.AddProduct(WexBimProduct.ReadFromStream(reader));
        }

        wexBimStream.Regions.forEach((region) => {
            let geometryCount = reader.readInt32();
            for (let i = 0; i < geometryCount; i++) {
                region.AddGeometryModel(WexBimGeometryModel.ReadFromStream(reader));
            }
        });

        //set up products with shapes
        wexBimStream.Regions.forEach((region) => {
            region.GeometryModels.forEach((model) => {
                model.Shapes.forEach((shape) => {
                    var product = wexBimStream.Products[shape.ProductLabel];
                    product.Shapes.push(shape);
                });
            });
        });

        return wexBimStream;
    }
}
